### PR TITLE
Subsriber fix

### PR DESCRIPTION
**Pull request by travipross**

### CODE REVIEW

**Comment by travipross** [comment]
Working on subscriber fixes here.

Based on the changes introduced by [PR #493 in the official repo](https://github.com/RobotWebTools/rosbridge_suite/pull/493).

File: rosbridge_library/src/rosbridge_library/internal/message_conversion.py
Comment (srubs) [inline rosbridge_library/src/rosbridge_library/internal/message_conversion.py:76]: Oh nice, you have also fixed arrays.

**Review by srubs — COMMENTED** [review]
I don't fully understand what is going on here. But it looks like it makes sense.

File: rosbridge_library/src/rosbridge_library/internal/message_conversion.py
Comment (travipross) [reply rosbridge_library/src/rosbridge_library/internal/message_conversion.py:76]: Can't take credit for this one, that was done by the upstream PR that's been stuck "in review"

File: rosbridge_library/src/rosbridge_library/internal/message_conversion.py
Comment (travipross) [inline rosbridge_library/src/rosbridge_library/internal/message_conversion.py:344]: @srubs out of curiousity, this is the part of the code that the reviewer upstream had seemed to object to.

If the upstream PR is stagnant much longer, I may re-submit a new PR that omits this shortcut step so it can get the conversation moving forward again.

Comment (srubs) [reply rosbridge_library/src/rosbridge_library/internal/message_conversion.py:344]: Yeah. I'm all for that. The more of the official release that we make use of the better.

File: rosbridge_library/src/rosbridge_library/internal/message_conversion.py
Comment (travipross) [inline rosbridge_library/src/rosbridge_library/internal/message_conversion.py:212]: This was the reason GPSFix messages were failing to parse.

For ROS2 fields, a dynamic array such as `int8[] my_arr` is defined in IDL  as `sequence<int8>`, for which the compiled regex `list_tokens`, defined above, will extract `int8` as the `rostype`.

However, IDL specifies a static array type such as `int8[3] my_arr` with a simple `int8[3]`. In this case, the `list_tokens` regex throws an `AttributeError`, so now I just catch it and try again with a different regex pattern.

See [here](https://design.ros2.org/articles/idl_interface_definition.html) for ROS2 IDL type definitions. There is a table near the bottom that describes array types. This page also gives some insight as to why additional type mappings were required to address past issues (e.g. `boolean` vs `bool`, etc.)

**Review by travipross — COMMENTED** [review]
The exceptions regarding message conversion have now been resolved, and we can now use `roslibpy` to subscribe to all the `hfss` topics as they're currently defined.

**Comment by travipross** [comment]
The `tornado.websocket.WebSocketClosedError` I've seen on the `rosbridge_webserver` node appears to be non-breaking, and likely just happening for large messages when a client disconnects mid-transmission.

What I've learned is that for any topic for which `rosbridge_suite` has a connected client subscribed to, each message is scheduled to be sent over the websocket to that client asynchronously via a `tornado.ioloop` callback. If a client (e.g. `roslibpy.Ros()`) suddenly disconnects, that subscriber is unregistered on the `rosbridge_suite` side and no new "message send" actions are registered to callbacks. However, it's possible that the callback for the most recent message on that topic has already been registered, and the message has therefore already been scheduled to be (or is already in the process of being) sent across the now-closed websocket, causing the above exception. 

This makes sense given that a larger message size increases the likelihood of disconnecting pre/mid-message. This was particularly noticeable in `Image` messages, as they are comparatively much larger in size than any other sample message we had tested. 

I think I've confirmed this theory, as I've modified my `roslibpy` subscription test code to first unsubscribe, then wait 1 second before closing the websocket. The error no longer occurs on the `rosbridge_webserver` node. This makes sense, as since the `rosbridge_webserver` unregisters the client when the client unsubscribes, no further messages are registered to be sent via callback. This gives the server time to send any remaining message over the socket before the connection closes, completely bypassing the issue.

Thankfully, even if the websocket were to close suddenly and unexpectedly from the client side, the `rosbridge_webserver` node raises the exception but continues to operate as intended, so all of this is really seeming like it won't be an issue.

**Comment by travipross** [comment]
I've done a bit of reading on this lately, and as it turns out you can't make a forked repository private, which means anything we add to this repo is open to the public. Nothing proprietary in here so far (and may never will be), but just something to keep in mind moving forward if we choose to start adding convenience scripts and whatnot to the repo.

The resolution to this in that case is to "duplicate" the official repository, which unlinks it from the forked base and allows us to set a copy as private. After we merge this PR, I may do some repo cleanup and change things around, as it'd be nice to have some circleCI scripts added here.

The nice thing is you can easily add multiple remotes to a local git repo, so we can keep this public one for official fixes (from others, or from us), while maintaining a private copy separately. We can just merge future improvements from the official repo into our private copy if we find they'd be useful.

**Comment by srubs** [comment]
> I've done a bit of reading on this lately, and as it turns out you can't make a forked repository private, which means anything we add to this repo is open to the public. Nothing proprietary in here so far (and may never will be), but just something to keep in mind moving forward if we choose to start adding convenience scripts and whatnot to the repo.
> 
> The resolution to this in that case is to "duplicate" the official repository, which unlinks it from the forked base and allows us to set a copy as private. After we merge this PR, I may do some repo cleanup and change things around, as it'd be nice to have some circleCI scripts added here.
> 
> The nice thing is you can easily add multiple remotes to a local git repo, so we can keep this public one for official fixes (from others, or from us), while maintaining a private copy separately. We can just merge future improvements from the official repo into our private copy if we find they'd be useful.

I don't mind this being a public repo as, in an ideal world, we would get this merged into the official repo.

**Review by srubs — APPROVED** [review]
I am happy to merge this in if you are ready for it to be merged.

**Comment by travipross** [comment]
> I don't mind this being a public repo as, in an ideal world, we would get this merged into the official repo.

Yes, that's fair. I was a bit torn on what to do when I initially wrote that comment. I guess I would just say that if we find it would make our lives easier to ever put in some custom code, or setup scripts, etc, then at least we would hypothetically have a means to do so without exposing anything meant to be internal.

I agree though, no need to change this as it is now. It makes sense to keep it public to both pull from and push to the upstream official one.

**Pull request merged**
